Sources/GRPCCore/Server.swift

```diff
@@ -0,0 +1,428 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import Atomics
+
+/// A gRPC server.
+///
+/// The server accepts connections from clients and listens on each connection for new streams
+/// which are initiated by the client. Each stream maps to a single RPC. The server routes accepted
+/// streams to a service to handle the RPC or rejects them with an appropriate error if no service
+/// can handle the RPC.
+///
+/// A ``Server`` may listen with multiple transports (for example, HTTP/2 and in-process) and route
+/// requests from each transport to the same service instance. You can also use "interceptors",
+/// to implement cross-cutting logic which apply to all accepted RPCs. Example uses of interceptors
+/// include request filtering, authentication, and logging. Once requests have been intercepted
+/// they are passed to a handler which in turn returns a response to send back to the client.
+///
+/// ## Creating and configuring a server
+///
+/// The following example demonstrates how to create and configure a server.
+///
+/// ```swift
+/// var server = Server()
+///
+/// // Create and add an in-process transport.
+/// let inProcessTransport = InProcessServerTransport()
+/// server.transports.add(inProcessTransport)
+///
+/// // Create and register the 'Greeter' and 'Echo' services.
+/// server.services.register(GreeterService())
+/// server.services.register(EchoService())
+///
+/// // Create and add some interceptors.
+/// server.interceptors.add(StatsRecordingServerInterceptors())
+/// ```
+///
+/// ## Starting and stopping the server
+///
+/// Once you have configured the server call ``run()`` to start it. Calling ``run()`` starts each
+/// of the servers transports. A ``ServerError`` is thrown if any of the transports can't be
+/// started.
+///
+/// ```swift
+/// // Start running the server.
+/// try await server.run()
+/// ```
+///
+/// The ``run()`` method won't return until the server has finished handling all requests. You can
+/// signal to the server that it should stop accepting new requests by calling ``stopListening()``.
+/// This allows the server to drain existing requests gracefully. To stop the server more abruptly
+/// you can cancel the task running your server. If your services require additional resources that
+/// need their lifecycles managed you should consider using [Swift Service
+/// Lifecycle](https://github.com/swift-server/swift-service-lifecycle).
```

Comment:
I love the callout but I think we should be even more specific here. Not only when you need handling additional resources but also when you want to add support to gracefully shutdown the server. WDYT?

Reply:
Can you be more specific? We support graceful shutdown here...

Reply:
Maybe we can give a general recommendation that we recommend running the `server` in a `ServiceGroup` since that gives you orchestration of multiple services and support for listening to signals to trigger graceful shutdown.

Reply:
I'm a bit hesitant to document specifics here. I think the best solution is to have a DocC article which runs through using lifecycle with grpc-swift (which we should link to from these docs). It also means there's a single place which needs to be kept up-to-date.